sympy/stats/stochastic_process_types.py

```diff
@@ -421,6 +421,15 @@ def _preprocess(self, given_condition, evaluate):
 
         return is_insufficient, trans_probs, state_index, given_condition
 
+    def replace_with_index(self, condition):
+        if isinstance(condition, Relational):
+            if condition.rhs.is_symbol:
+                try:
+                    return Eq(condition.lhs, self.state_space.index(condition.rhs))
+                except NameError:
+                    raise NameError('%s not found in the state space'%condition.rhs)
```

Comment:
```suggestion
        if isinstance(condition, Relational):
            return type(condition)(self.state_space.index(condition.lhs), self.state_space.index(condition.rhs))
```

Reply:
I also see that, the implementation of `index_of` in `master` is a bit inefficient,

https://github.com/sympy/sympy/blob/bdb49c4abfb35554a3c8ce761696ffff3bb837fe/sympy/stats/stochastic_process_types.py#L753-L766

The dictionary should be computed only once and stored as one of the attributes of the object. Can you please optimise it in your PR?

Reply:
Yeah, I'll fix that. I was wondering if `self.state_space.index_of[condition.lhs]` is required as lhs would remain the same.

Reply:
Lastly, for strings, right now, its possible to pass Eq(Y[3],Str('Cloudy')) and we can convert this to index in replace_with_index. Would that be okay?

Reply:
> I was wondering if `self.state_space.index_of[condition.lhs]` is required as lhs would remain the same.

What about, `Gt(2, Y[3])` or `Eq(Str('Cloudy'), Y[3])` or in fact, how does this work, `Eq(Y[5], Y[6])`?

Reply:
> What about, `Gt(2, Y[3])` or `Eq(Str('Cloudy'), Y[3])` or in fact, how does this work, `Eq(Y`

Oh, I'll add an `isinstance(Y[i],RandomIndexedSymbol)` check to make sure I replace the right side of the equation